include/deal.II/multigrid/multigrid.h

```diff
@@ -74,6 +74,60 @@ class Multigrid : public Subscriptor
     f_cycle
   };
 
+  /**
+   * A structure that has boost::signal objects for a number of actions that
+   * happen for the typical usage of this class.
+   *
+   * For documentation on signals, see
+   * http://www.boost.org/doc/libs/release/libs/signals2 .
+   */
+  struct Signals
+  {
+    /**
+     * This signal is triggered before and after the call to
+     * MGTransfer::copy_to_mg which transfers given vector to all levels.
+     */
+    boost::signals2::signal<void (bool start)> transfer_to_mg;
+
+    /**
+     * This signal is triggered before and after the call to
+     * MGTransfer::copy_from_mg which transfers a multi-level vector
+     * to a normal vector.
+     */
+    boost::signals2::signal<void (bool start)> transfer_to_global;
+
+    /**
+     * This signal is triggered before and after the call to the coarse solver.
+     */
+    boost::signals2::signal<void (const bool start, const unsigned int level)> coarse_solve;
+
+    /**
+     * This signal is triggered before and after the call to
+     * MGTransfer::restrict_and_add() which restricts a vector from @p level
+     * to the next coarser one (@p level - 1).
+     */
+    boost::signals2::signal<void (const bool start, const unsigned int level)> transfer_down_to;
+
+    /**
+     * This signal is triggered before and after the call to
+     * MGTransfer::prolongate() which prolongates a vector to @p level
+     * from the next coarser one (@p level - 1).
+     */
+    boost::signals2::signal<void (const bool start, const unsigned int level)> transfer_up_to;
+
+    /**
+     * This signal is triggered before and after the call to
+     * a pre-smoothing step via MGPreSmoother::apply().
+     */
+    boost::signals2::signal<void (const bool start, const unsigned int level)> pre_smoother_step;
```

Comment:
Here and below: document the arguments.

Are pre- and post-smoothers usually different? Is it worth having two separate signals for this, or would it be easier in practice to merge the two signals and given the resulting one another bool argument that denotes pre/post smoothing?

Reply:
No, normally we use the same object.
Following the concept of this PR to have different signals for different events, I think that having two separate signals is appropriate nevertheless. 
At least, we should treat them the same as `transfer_up_to` and `transfer_down_to`.

Reply:
That sounds fine to me.